### PR TITLE
Make `CXMappingPass` a `StandardPass` with round-trip serialization and deserialization

### LINKED ISSUE #1573: Make `CXMappingPass()` serializable [was: Replace the `RebaseCustom` in `qiskit_convert.py` with a serializable pass]

### PR DESCRIPTION
**Pull request by cqc-alec**
# Description

Use `AutoRebase` instead of `RebaseCustom` in the pass; and wrap up the `SequencePass` in a `StandardPass` so that serialization and deserialization are inverse.

# Related issues

Fixes #1573 .

# Checklist

- [x] I have performed a self-review of my code.
- [ ] I have commented hard-to-understand parts of my code.
- [ ] I have made corresponding changes to the public API documentation.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have updated the changelog with any user-facing changes.

### CODE REVIEW

File: tket/src/Predicates/PassGenerators.cpp [hunk 1]
```diff
@@ -480,14 +481,28 @@ PassPtr gen_cx_mapping_pass(
...
-  PassPtr rebase_pass =
-      gen_rebase_pass(gate_set, CircPool::CX(), CircPool::tk1_to_tk1);
+  PassPtr rebase_pass = gen_auto_rebase_pass(gate_set);
   PassPtr return_pass =
       rebase_pass >> gen_full_mapping_pass(arc, placement_ptr, config);
   if (delay_measures) return_pass = return_pass >> DelayMeasures();
   return_pass = return_pass >> rebase_pass >>
                 gen_decompose_routing_gates_to_cxs_pass(arc, directed_cx);
-  return return_pass;
+  Transform t{[=](Circuit& circ) {
```
Comment (cqc-alec) [inline tket/src/Predicates/PassGenerators.cpp:490]: I am not totally confident in this: should we be passing a `std::shared_ptr<unit_bimaps_t>` as well as a `Circuit&` to the `Transform` constructor? If so what should we do with it?

Comment (sjdilkes) [reply tket/src/Predicates/PassGenerators.cpp:490]: What would the `unit_bimaps_t` object be for? I might be forgetting some TKET infrastructure here. Certainly we don't need to pass any kind of qubit placement if that's what it's referring to (as the circuit UnitID are updated).

**Review by sjdilkes — CHANGES_REQUESTED** [review]
Looks fine to me, but I've asked for some clarification on the question.

File: tket/src/Predicates/PassGenerators.cpp (hunk 1 above)
Comment (cqc-alec) [reply tket/src/Predicates/PassGenerators.cpp:490]: Wasn't sure myself. Seems to be used in placement passes where units are renamed (e.g. `gen_placement_pass()`). So I assume not necessary here.

**Review by sjdilkes — CHANGES_REQUESTED** [review]

File: tket/src/Predicates/PassGenerators.cpp (hunk 1 above)
Comment (cqc-alec) [reply tket/src/Predicates/PassGenerators.cpp]: Please check my last commit. I don't know if this is correct, but I don't know what else to do with the maps either!

File: tket/src/Predicates/PassGenerators.cpp [hunk 2]
```diff
@@ -480,14 +481,29 @@ PassPtr gen_cx_mapping_pass(
...
-  PassPtr rebase_pass =
-      gen_rebase_pass(gate_set, CircPool::CX(), CircPool::tk1_to_tk1);
+  PassPtr rebase_pass = gen_auto_rebase_pass(gate_set);
   PassPtr return_pass =
       rebase_pass >> gen_full_mapping_pass(arc, placement_ptr, config);
   if (delay_measures) return_pass = return_pass >> DelayMeasures();
   return_pass = return_pass >> rebase_pass >>
                 gen_decompose_routing_gates_to_cxs_pass(arc, directed_cx);
-  return return_pass;
+  Transform t{[=](Circuit& circ, std::shared_ptr<unit_bimaps_t> maps) {
```
Comment (sjdilkes) [inline tket/src/Predicates/PassGenerators.cpp]: Does this work?
`Transform t{[=](Circuit& circ, std::shared_ptr<unit_bimaps_t> maps) {

      // Relabel all two-qubit gates
      CompilationUnit cu(circ);
      bool changed = rebase_pass->apply(cu);
      circ = cu.get_circ_ref();
      // Now try placement, falling back on "LinePlacement" if "GraphPlacement" fails
      try {
        changed = changed | placement_ptr->place(circ, maps);
      } catch (const std::runtime_error& e) {
        std::stringstream ss;
        ss << "PlacementPass failed with message: " << e.what()
          << " Fall back to LinePlacement.";
        tket_log()->warn(ss.str());
        Placement::Ptr line_placement_ptr = std::make_shared<LinePlacement>(
            placement_ptr->get_architecture_ref());
        changed = changed | line_placement_ptr->place(circ, maps);

      }
      // Now route
      MappingManager mm(std::make_shared<Architecture>(arc));
      changed = changed | mm.route_circuit_with_maps(circ, config, maps);
      
      // Now decompose routing gates to cx gates (this won't change maps)
      cu(circ);
      if(delay_measures){
        DelayMeasures()->apply(cu);
      }
      gen_decompose_routing_gates_to_cxs_pass(arc, directed_cx)->apply(cu);
      circ = cu.get_circ_ref();
      return changed;
  }`

File: tket/src/Predicates/PassGenerators.cpp (hunk 1 above)
Comment (sjdilkes) [reply tket/src/Predicates/PassGenerators.cpp:490]:
```  Transform t{[=](Circuit& circ, std::shared_ptr<unit_bimaps_t> maps) {

      // Relabel all two-qubit gates
      CompilationUnit cu(circ);
      bool changed = rebase_pass->apply(cu);
      circ = cu.get_circ_ref();
      // Now try placement, falling back on "LinePlacement" if "GraphPlacement" fails
      try {
        changed = changed | placement_ptr->place(circ, maps);
      } catch (const std::runtime_error& e) {
        std::stringstream ss;
        ss << "PlacementPass failed with message: " << e.what()
          << " Fall back to LinePlacement.";
        tket_log()->warn(ss.str());
        Placement::Ptr line_placement_ptr = std::make_shared<LinePlacement>(
            placement_ptr->get_architecture_ref());
        changed = changed | line_placement_ptr->place(circ, maps);

      }
      // Now route
      MappingManager mm(std::make_shared<Architecture>(arc));
      changed = changed | mm.route_circuit_with_maps(circ, config, maps);
      
      // Now decompose routing gates to cx gates (this won't change maps)
      cu(circ);
      if(delay_measures){
        DelayMeasures()->apply(cu);
      }
      gen_decompose_routing_gates_to_cxs_pass(arc, directed_cx)->apply(cu);
      circ = cu.get_circ_ref();
      return changed;
  }```

**Review by sjdilkes — CHANGES_REQUESTED** [review]

**Review by sjdilkes — APPROVED** [review]

File: tket/src/Predicates/PassGenerators.cpp [hunk 3]
```diff
@@ -480,14 +481,53 @@ PassPtr gen_cx_mapping_pass(
     bool delay_measures) {
   OpTypeSet gate_set = all_single_qubit_types();
   gate_set.insert(OpType::CX);
-  PassPtr rebase_pass =
-      gen_rebase_pass(gate_set, CircPool::CX(), CircPool::tk1_to_tk1);
+  PassPtr rebase_pass = gen_auto_rebase_pass(gate_set);
   PassPtr return_pass =
```
Comment (sjdilkes) [inline tket/src/Predicates/PassGenerators.cpp:485]: We can remove the references to return pass here I think

**Review by sjdilkes — CHANGES_REQUESTED** [review]

File: tket/src/Predicates/PassGenerators.cpp (hunk 3 above)
Comment (cqc-alec) [reply tket/src/Predicates/PassGenerators.cpp:485]: It's used below in the line
```
  PassConditions conditions = return_pass->get_conditions();
```

Comment (cqc-alec) [reply tket/src/Predicates/PassGenerators.cpp:485]: ...which I think is still correct, although there may be an easier way to construct these conditions.

**Review by sjdilkes — APPROVED** [review]

**Pull request merged**